server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuthentication.scala

```diff
@@ -0,0 +1,192 @@
+
+    val nonce = params("nonce")
+    val nc = params("nc")
+    val f = ask(nonce_keeper, NonceKeeper.ReceiveNonce(nonce, Integer.parseInt(nc, 16)))(ATimeout(5.seconds)).mapTo[NonceKeeper.Reply]
+    Await.result(f, 5.seconds) match {
+      case NonceKeeper.StaleReply => StaleNonce
+      case NonceKeeper.BadNCReply => BadNC
+      case NonceKeeper.OKReply =>
+        if (!store.isDefinedAt((realm, params("username"))))
+          UserUnknown
+        else
+          store((realm, params("username"))) match {
+            case password => {
+              val resp = DigestUtil.computeResponse(method, params("username"), realm, password, uri, nonce, nc, params("cnonce"), params("qop"))
+              if (resp == params("response"))
+                OK
+              else
+                WrongResponse
+            }
+          }
+    }
+  }
+}
+
+class Nonce(val created: Date, var nc: Int, val data: String) {
+}
+
+object Nonce {
+  val random = new SecureRandom()
+
+  private def getRandomData(bits: Int) = new BigInteger(bits, random).toString(16)
+
+  def apply(bits: Int) = {
+    new Nonce(new Date(), 0, getRandomData(bits))
+  }
+}
+
+object NonceKeeper {
+
+  case object NewNonce
+
+  case class ReceiveNonce(data: String, nc: Int)
+
+  case object RemoveStale
+
+  abstract class Reply
+
+  case object StaleReply extends Reply
+
+  case object OKReply extends Reply
+
+  case object BadNCReply extends Reply
+
+}
+
+/**
+ * An {@link Actor} used to manage a database of nonces.
+ * Sending the {@link NonceKeeper.NewNonce} message yields a reply
+ * containing a fresh nonce as a {@link String}. On receiving a
+ * {@link NonceKeeper.ReceiveNonce} message, this actor checks if it
+ * is known and the nc value is correct. If so, the nc value associated
+ * with the nonce is increased and either {@link NonceKeeper.OKReply}
+ * or {@link NonceKeeper.BadNCReply} is replied. Finally, on
+ * receiving a {@link NonceKeeper.RemoveStale} message, the actor
+ * removes nonces that are older than staleTimeout.
+ */
+class NonceKeeper(staleTimeout: Long, bits: Int) extends Actor {
+  assert(bits > 0)
+  var nonces: Map[String, Nonce] = new HashMap[String, Nonce]
+
+  def isStale(past: Date, now: Date) = staleTimeout < now.getTime() - past.getTime()
+
+  def receive = {
+    case NonceKeeper.NewNonce => {
+      var n: Nonce = null
+      do {
+        n = Nonce(bits)
+      } while (nonces.contains(n.data))
+      nonces = nonces + (n.data -> n)
+      sender() ! n.data
+    }
+    case NonceKeeper.ReceiveNonce(data, nc) => {
+      nonces.get(data) match {
+        case None => sender() ! NonceKeeper.StaleReply
+        case Some(n) => {
+          if (nc > n.nc) {
+            n.nc = n.nc + 1
+            sender() ! NonceKeeper.OKReply
+          } else
+            sender() ! NonceKeeper.BadNCReply
+        }
+      }
+    }
+    case NonceKeeper.RemoveStale =>
+      nonces = nonces.filter { case (_, n) =>
+        !isStale(n.created, new Date())
```

Comment:
We can make a single new date outside the filter call so we are both uniform in the filter and reduce allocations.